src/cmdline/mod.rs

```diff
@@ -90,117 +95,294 @@ fn valid_element(s: &str) -> Result<()> {
+    /// assert_eq!(cl.set_capacity(3), Err(Error::InvalidCapacity));
+    /// ```
+    pub fn set_capacity(&mut self, new_capacity: usize) -> Result<()> {
+        if self.get_null_terminated_representation_size() <= new_capacity {
+            self.capacity = new_capacity;
+
             Ok(())
         } else {
-            Err(Error::TooLarge)
+            Err(Error::InvalidCapacity)
         }
     }
 
-    fn start_push(&mut self) {
-        if !self.line.is_empty() {
-            self.line.push(' ');
+    fn get_null_terminated_representation_size(&self) -> usize {
+        // Computing current size of the cmdline
+        let mut cmdline_size = self.boot_args.len() + 1; // for null terminator
+
+        if !self.init_args.is_empty() {
+            cmdline_size += INIT_ARGS_SEPARATOR.len() + self.init_args.len();
         }
+
+        cmdline_size
     }
 
-    fn end_push(&mut self) {
-        // This assert is always true because of the `has_capacity` check that each insert method
-        // uses.
-        assert!(self.line.len() < self.capacity);
+    fn has_capacity(&self, more: usize, args_type: CmdlineArgType) -> bool {
+        let mut cmdline_size = self.get_null_terminated_representation_size();
+
+        // Adding extra size required for the insertion of `more` bytes as a `args_type` argument
+
+        // Checking if one extra space needs to be added before insertion of a new argument
+        cmdline_size += match args_type {
+            CmdlineArgType::BootArg => {
+                if !self.boot_args.is_empty() {
+                    1
+                } else {
+                    0
+                }
+            }
+            CmdlineArgType::InitArg => {
+                if !self.init_args.is_empty() {
+                    1
+                } else {
+                    INIT_ARGS_SEPARATOR.len()
+                }
+            }
+        };
+
+        cmdline_size += more;
+
+        cmdline_size <= self.capacity
     }
 
-    /// Validates and inserts a key-value pair into this command line.
+    /// Validates and inserts a key=value pair as a new cmdline argument.
     ///
     /// # Arguments
     ///
-    /// * `key` - Key to be inserted in the command line string.
+    /// * `key` - Key to be inserted in the args string.
     /// * `val` - Value corresponding to `key`.
+    /// * `args_type` - Type of the new argument to be inserted.
     ///
     /// # Examples
     ///
     /// ```rust
     /// # use linux_loader::cmdline::*;
-    /// let mut cl = Cmdline::new(100);
-    /// cl.insert("foo", "bar");
+    /// let mut cl = Cmdline::new(10).unwrap();
+    /// cl.insert_key_value("foo", "bar", CmdlineArgType::BootArg)
+    ///     .unwrap();
     /// assert_eq!(cl.as_cstring().unwrap().as_bytes_with_nul(), b"foo=bar\0");
     /// ```
-    pub fn insert<T: AsRef<str>>(&mut self, key: T, val: T) -> Result<()> {
+    pub fn insert_key_value<T: AsRef<str>>(
```

Comment:
Why not leave it as insert? The fact that it's key-value is clear from the parameters that it receives.

Reply:
going back to the old naming